File: app/src/main/java/me/srikavin/quiz/viewmodel/GameViewModel.java
package me.srikavin.quiz.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import me.srikavin.quiz.model.AnswerResponse;
import me.srikavin.quiz.model.Quiz;
import me.srikavin.quiz.model.QuizAnswer;
import me.srikavin.quiz.model.QuizGameState;
import me.srikavin.quiz.model.QuizQuestion;
import me.srikavin.quiz.repository.GameRepository;
import me.srikavin.quiz.repository.QuizRepository;

public class GameViewModel extends ViewModel {
    private MutableLiveData<Quiz> quiz;

    private MutableLiveData<Integer> gameScore = new MutableLiveData<>();

    private MutableLiveData<String> currentGameID = new MutableLiveData<>();
    private MutableLiveData<GameRepository.GameInfo> gameInfo = new MutableLiveData<>();
    private MutableLiveData<QuizQuestion> currentQuestion = new MutableLiveData<>();
    private MutableLiveData<AnswerResponse> answerResponse = new MutableLiveData<>();
    private MutableLiveData<GameRepository.GameStats> gameStats = new MutableLiveData<>();
    private MutableLiveData<QuizGameState> gameState = new MutableLiveData<>();
    private MutableLiveData<Integer> timeRemaining = new MutableLiveData<>();
    private MutableLiveData<Void> createGame;


    public LiveData<String> getCurrentGameID() {
        return currentGameID;
    }

    public LiveData<GameRepository.GameInfo> getGameInfo() {
        return gameInfo;
    }

    public LiveData<QuizGameState> getGameState() {
        return gameState;
    }

    public LiveData<Integer> getTimeRemaining() {
        return timeRemaining;
    }

    public LiveData<QuizQuestion> getCurrentQuestion() {
        return currentQuestion;
    }

    public LiveData<AnswerResponse> getAnswerResponse() {
        return answerResponse;
    }

    public LiveData<GameRepository.GameStats> getGameStats() {
        return gameStats;
    }

    public LiveData<Integer> getGameScore() {
        return gameScore;
    }

    public LiveData<Quiz> getQuizByID(String id) {
        if (quiz == null) {
            quiz = new MutableLiveData<>();
            loadQuizzes(id);
        }
        return quiz;
    }

    public void createGame(Quiz quiz) {
        if (createGame == null) {
            createGame = new MutableLiveData<>();
            GameRepository.INSTANCE.createGame(quiz, new GameRepository.GameResponseHandler() {
                @Override
                public void handleAnswer(AnswerResponse response) {
                    answerResponse.postValue(response);
                }

                @Override
                public void handleQuestion(QuizQuestion question) {
                    currentQuestion.postValue(question);
                }

                @Override
                public void handleGameStateChange(QuizGameState state) {
                    gameState.postValue(state);
                }

                @Override
                public void handleScoreChange(int score) {
                    gameScore.postValue(score);
                }

                @Override
                public void handleGameCreate(String id, GameRepository.GameInfo info) {
                    currentGameID.postValue(id);
                    gameInfo.postValue(info);
                }

                @Override
                public void handleGameStats(GameRepository.GameStats stats) {
                    gameStats.postValue(stats);
                }

                @Override
                public void handleGameTimeChange(int timeLeft) {
                    timeRemaining.postValue(timeLeft);

                }
            });
        }
        return;
    }

    public void submitAnswer(QuizAnswer quizAnswer) {
        GameRepository.INSTANCE.submitAnswer(currentGameID.getValue(), quizAnswer);
    }

    private void loadQuizzes(String id) {
        QuizRepository.INSTANCE.getQuizByID(id, new QuizRepository.QuizResponseHandler() {
            @Override
            public void handle(Quiz quiz) {
                GameViewModel.this.quiz.postValue(quiz);
            }
        });
    }
}

File: app/src/main/java/me/srikavin/quiz/viewmodel/GameViewModel.kt
package me.srikavin.quiz.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.ww.roxie.BaseAction
import com.ww.roxie.BaseState
import me.srikavin.quiz.model.AnswerResponse
import me.srikavin.quiz.model.Quiz
import me.srikavin.quiz.model.QuizGameState
import me.srikavin.quiz.network.common.model.data.QuizAnswerModel
import me.srikavin.quiz.network.common.model.data.QuizQuestionModel
import me.srikavin.quiz.repository.GameID
import me.srikavin.quiz.repository.GameRepository
import me.srikavin.quiz.repository.QuizRepository

sealed class GameAction : BaseAction {
    data class LoadQuiz(val quizId: String) : GameAction()
}

sealed class GameChange {
    object Loading : GameChange()
    data class QuizLoad(val quiz: Quiz) : GameChange()
    data class Error(val error: Throwable?) : GameChange()
}

data class GameViewState(
        val quiz: Quiz? = null,
        val loading: Boolean = false,
        val error: Boolean = false
) : BaseState

class GameViewModel(application: Application) : AndroidViewModel(application) {

    private val quizRepository: QuizRepository = QuizRepository(application)

    private var quiz = MutableLiveData<Quiz>()

    private val gameScore = MutableLiveData<Int>()

    private val currentGameID = MutableLiveData<GameID>()
    private val gameInfo = MutableLiveData<GameRepository.GameInfo>()
    private val currentQuestion = MutableLiveData<QuizQuestionModel>()
    private val answerResponse = MutableLiveData<AnswerResponse>()
    private val gameStats = MutableLiveData<GameRepository.GameStats>()
    private val gameState = MutableLiveData<QuizGameState>()
    private val timeRemaining = MutableLiveData<Int>()
    private var createGame: MutableLiveData<Void>? = null


    fun getCurrentGameID(): LiveData<GameID> {
        return currentGameID
    }

    fun getGameInfo(): LiveData<GameRepository.GameInfo> {
        return gameInfo
    }

    fun getGameState(): LiveData<QuizGameState> {
        return gameState
    }

    fun getTimeRemaining(): LiveData<Int> {
        return timeRemaining
    }

    fun getCurrentQuestion(): LiveData<QuizQuestionModel> {
        return currentQuestion
    }

    fun getAnswerResponse(): LiveData<AnswerResponse> {
        return answerResponse
    }

    fun getGameStats(): LiveData<GameRepository.GameStats> {
        return gameStats
    }

    fun getGameScore(): LiveData<Int> {
        return gameScore
    }

    fun getQuizByID(id: String): LiveData<Quiz> {
        quiz = MutableLiveData()
        loadQuizzes(id)
        return quiz
    }

    fun stopMatchmaking() {
        GameRepository.stopMatchmaking()
    }

    fun createGame(quiz: Quiz, remote: Boolean) {
        if (createGame == null) {
            createGame = MutableLiveData()
            GameRepository.createGame(quiz, object : GameRepository.GameResponseHandler() {
                override fun handleAnswer(response: AnswerResponse) {
                    answerResponse.postValue(response)
                }

                override fun handleQuestion(question: QuizQuestionModel) {
                    currentQuestion.postValue(question)
                }

                override fun handleGameStateChange(state: QuizGameState) {
                    gameState.postValue(state)
                }

                override fun handleScoreChange(score: Int) {
                    gameScore.postValue(score)
                }

                override fun handleGameInfo(info: GameRepository.GameInfo) {
                    gameInfo.postValue(info)
                }

                override fun handleGameCreate(id: GameID) {
                    currentGameID.postValue(id)
                }

                override fun handleGameStats(stats: GameRepository.GameStats) {
                    gameStats.postValue(stats)
                }

                override fun handleGameTimeChange(timeLeft: Int) {
                    timeRemaining.postValue(timeLeft)

                }
            }, remote)
        }
        return
    }

    fun submitAnswer(quizAnswer: QuizAnswerModel) {
        GameRepository.submitAnswer(currentGameID.value
                ?: throw RuntimeException("Current game ID is null; cannot call submit answer here"), quizAnswer)
    }

    fun quitGame() {
        GameRepository.quit(currentGameID.value ?: return)
    }

    private fun loadQuizzes(id: String) {
        quizRepository.getQuizByID(id, object : QuizRepository.QuizResponseHandler() {
            override fun handle(quiz: Quiz?) {
                this@GameViewModel.quiz.postValue(quiz)
            }
        })
    }
}
